src/Language/Spectacle/Syntax/Logic.hs

```diff
@@ -0,0 +1,159 @@
+  Lang ctx effs a
+exists xs p = scope (Exists (foldMap (: []) xs) p)
+{-# INLINE exists #-}
+
+-- | Logical negation. The 'complement' operator is equivalent to 'not' for simple expressions, but
+-- can be used to negate quantifiers and the other logical operators in spectacle.
+--
+-- @since 0.1.0.0
+complement :: Member Logic effs => Lang ctx effs a -> Lang ctx effs a
+complement m = scope (Complement m)
+
+-- | Boolean conjunction.
+--
+-- @since 0.1.0.0
+conjunct :: Member Logic effs => Lang ctx effs Bool -> Lang ctx effs Bool -> Lang ctx effs Bool
+conjunct m n = scope (Conjunct m n)
+
+-- | Boolean disjunction.
+--
+-- @since 0.1.0.0
+disjunct :: Member Logic effs => Lang ctx effs Bool -> Lang ctx effs Bool -> Lang ctx effs Bool
+disjunct m n = scope (Disjunct m n)
+
+-- | Logical implication.
+--
+-- @since 0.1.0.0
+implies :: Member Logic effs => Lang ctx effs Bool -> Lang ctx effs Bool -> Lang ctx effs Bool
+implies m n = complement (conjunct m (complement n))
+
+-- | If and only if.
+--
+-- @since 0.1.0.0
+iff :: Member Logic effs => Lang ctx effs Bool -> Lang ctx effs Bool -> Lang ctx effs Bool
+iff m n = conjunct (m `implies` n) (n `implies` m)
+
+-- | Discharge a 'Logic' effect.
+--
+-- @since 0.1.0.0
+runLogic ::
+  Members '[Error RuntimeException, NonDet] effs =>
+  Lang ctx (Logic ': effs) a ->
+  Lang ctx effs a
+runLogic = \case
+  Pure x -> Pure x
+  Yield (Op op) k -> case decomposeOp op of
+    Left other -> Yield (Op other) (runLogic . k)
+    Right bottom -> absurd (coerce bottom)
+  Yield (Scoped scoped loom) k -> case decomposeS scoped of
+    Left other -> Yield (Scoped other loom') k'
+    Right (Forall xs p) ->
+      k' =<< runLoom loom' do
+        b <- fmap and (mapM p xs)
+        unless b (throwE (QuantifierException ForallViolated))
+        oneOf xs
+    Right (Exists xs p) ->
+      k' =<< runLoom loom' do
+        xs' <- filterM p xs
+        when (null xs') (throwE (QuantifierException ExistsViolated))
+        oneOf xs'
+    Right (Complement m) -> runLoom loom' (rewriteComplement m) >>= k'
+    Right (Conjunct m n) -> runLoom loom' (liftA2 (&&) m n) >>= k'
+    Right (Disjunct m n) -> runLoom loom' (liftA2 (||) m n) >>= k'
+    where
+      loom' = loom ~>~ hoist runLogic
+
+      k' = runLogic . k
+
+-- | Applies negation to logical operators.
+--
+-- @since 0.1.0.0
+rewriteComplement :: Member Logic effs => Lang ctx effs a -> Lang ctx effs a
+rewriteComplement = \case
+  Pure x -> pure x
+  Yield (Op op) k -> Yield (Op op) (rewriteComplement . k)
+  Yield (Scoped scoped loom) k -> case projectS scoped of
+    Nothing -> Yield (Scoped scoped loom') k'
+    Just (Forall xs p) -> runLoom loom (exists xs (fmap not . p)) >>= k'
+    Just (Exists xs p) -> runLoom loom (forall xs (fmap not . p)) >>= k'
+    Just (Complement m) -> runLoom loom' m >>= k'
+    Just (Conjunct m n) -> do
+      let m' = complement (fmap not m)
+          n' = complement (fmap not n)
+      k' =<< runLoom loom' (disjunct m' n')
```

Comment:
Nit: inconsistent use of `>>=` and `=<<`. Perhaps just stick with `>>=` since you use it in the above cases too?